### ISSUE TITLE
VT102Writer.__str__ method in manhole.py does not check for special case

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11096]
|<img alt="htj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| htj reported|
|-|-|
|Trac ID|trac#1565|
|Type|defect|
|Created|2006-03-13 14:33:29Z|

From twisted.conch.manhole.py (around line 249)

```
    def __str__(self):
        s = ''.join(self.written)
        return s.strip('\n').splitlines()[-1]
```
The last line is bad, i.e., try doing:

```
>>> ''.splitlines()[-1]
Traceback (most recent call last):
  File "<stdin>", line 1, in ?
IndexError: list index out of range
>>> 'a'.splitlines()[-1]
'a'
```

The contents of the string needs to be checked before doing splitlines (or at some other time). I've cooked up a short patch. Attaching.

I can usually provoke this bug by connecting to a manhole server (used ssh) and sending Control+L. Expected behaviour was that this should clear the screen, but it does not. Not sure thats a bug though.

Attachments:

* [manhole_patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c04/c04038a9efd402b01f5733a5df5e0248d29a1ad8/2b12c9c25f6771e1372faf5a35117e2e498a1380.diff) (518 bytes) - added by htj on 2006-03-13 14:34:43Z - Patch that fixes the described problem.

<details><summary>Searchable metadata</summary>

```
trac-id__1565 1565
type__defect defect
reporter__htj htj
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1142260409000000 1142260409000000
changetime__1150428902000000 1150428902000000
version__ 
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

I've fixed this by having manhole actually handle ^L.  VT102Writer shouldn't have to deal with empty inputs, and handling ^L prevents it from having to do so.

I added a test for ^L and some things that are like tests for VT102Writer, but they don't actually assert anything (other than that no exceptions are raised).

Review in [branches/manhole-formfeed-1565](branches/manhole-formfeed-1565)

Suggestions for improving the VT102Writer tests welcome.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jerub|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @exarkun|
|-|-|

Fix the pyflakes problems as reasonable, manhole.py is the only one that's worth bothering about.
```
$ pyflakes test/test_helper.py test/test_manhole.py recvline.py insults/helper.py manhole.py
test/test_helper.py:6: 'ServerProtocol' imported but unused
test/test_helper.py:6: 'ClientProtocol' imported but unused
test/test_manhole.py:7: 'time' imported but unused
manhole.py:23: 'log' imported but unused
```

test/test_helper.py has had whitespace changes, please revert:
```
Index: test/test_helper.py
===================================================================
--- test/test_helper.py (revision 17236)
+++ test/test_helper.py (working copy)
@@ -363,6 +363,8 @@
             s3 + '\n' +
             '\n' * (HEIGHT - 3))
 
+
+
 class FakeDelayedCall:
     called = False
     cancelled = False
```

There are tests, and the tests pass, and this makes manhole saner. Thankyou for your attention to detail, and after you've changed the trivial things noted, please merge.

As for how to improve the tests, I've not got a suggestion.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [17246]) Merge manhole-formfeed-1565

Author: exarkun
Reviewer: jerub
Fixes [#11096](https://github.com/twisted/twisted/issues/11096)

Adds support for ^L to Manhole.  The behavior emulates CPython's interactive
mode.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
